module/zfs/zfs_vnops.c

```diff
@@ -932,19 +932,29 @@ zfs_write(struct inode *ip, uio_t *uio, int ioflag, cred_t *cr)
 }
 EXPORT_SYMBOL(zfs_write);
 
+/*
+ * Drop a reference on the passed inode asynchronously. This ensures
+ * that the caller will never drop the last reference on an inode in
+ * the current context. Doing so while holding open a tx could result
+ * in a deadlock if iput_final() re-enters the filesystem code.
+ */
 void
 zfs_iput_async(struct inode *ip)
 {
 	objset_t *os = ITOZSB(ip)->z_os;
+	taskqid_t id = 0;
 
 	ASSERT(atomic_read(&ip->i_count) > 0);
 	ASSERT(os != NULL);
 
-	if (atomic_read(&ip->i_count) == 1)
-		taskq_dispatch(dsl_pool_iput_taskq(dmu_objset_pool(os)),
-		    (task_func_t *)iput, ip, TQ_SLEEP);
-	else
+	if (atomic_read(&ip->i_count) == 1) {
+		while (id == 0) {
+			id = taskq_dispatch(dsl_pool_iput_taskq(dmu_objset_pool(os)),
+			    (task_func_t *)iput, ip, TQ_SLEEP);
+		}
```

Comment:
@tuxoko could you comment on this.  The issue here is that if `taskq_dispatch()` does ever fail we're going to leak an iput an inode.  Now because on the way taskq_dispatch is currently implemented this can't happen.  How do you feel about handling this impossible case by retrying vs simply adding a VERIFY or ASSERT that this can't happen.